File: typescript/test/conformanceScenario.test.ts
import { readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";
import {
  compareAnswer,
  parsePageJson,
  runCase,
  type PageDefinition,
  type ScenarioCase,
} from "../src/index.js";

// シナリオ（定義を動かして答えを見る）の共有フィクスチャ。Dart 版も同じものを食べる
// ＝**画面の中と道具の答えがズレない**ことが、この道具の値打ちなので。
const fixture = JSON.parse(
  readFileSync("../spec/conformance/scenario.json", "utf8"),
);

describe("conformance: scenario", () => {
  // strict で読む＝フィクスチャが「本当に書ける定義」であることも縛る。
  const page = parsePageJson(JSON.stringify(fixture.page), {
    strict: true,
  }) as PageDefinition;

  for (const one of fixture.cases as ScenarioCase[]) {
    it(one.name, () => {
      const answer = runCase(page, one);
      expect(compareAnswer(one.expect, answer)).toEqual([]);
    });
  }

  // 期待の**全部**が答えと一致することも見る（`compareAnswer` は書いた欄だけ見るので、
  // 表の側が欄を書き忘れても通ってしまう。表は6つの欄を全部持っている）。
  it("表はどの件でも5つの欄を全部書いている（書き忘れで緩くならない）", () => {
    for (const one of fixture.cases as ScenarioCase[]) {
      expect(Object.keys(one.expect ?? {}).sort(), one.name).toEqual([
        "computed",
        "enabled",
        "errors",
        "hidden",
        "required",
      ]);
    }
  });

  it("答えの中身そのものが表と同じ（順不同ではなく、まるごと）", () => {
    for (const one of fixture.cases as ScenarioCase[]) {
      const answer = runCase(page, one);
      expect(
        {
          errors: [...answer.errors].sort((a, b) =>
            `${a.field}${a.message}`.localeCompare(`${b.field}${b.message}`),
          ),
          computed: answer.computed,
          enabled: answer.enabled,
          hidden: [...answer.hidden].sort(),
          required: [...answer.required].sort(),
        },
        one.name,
      ).toEqual({
        errors: [...(one.expect?.errors ?? [])].sort((a, b) =>
          `${a.field}${a.message}`.localeCompare(`${b.field}${b.message}`),
        ),
        computed: one.expect?.computed,
        enabled: one.expect?.enabled,
        hidden: [...(one.expect?.hidden ?? [])].sort(),
        required: [...(one.expect?.required ?? [])].sort(),
      });
    }
  });
});

File: flutter/packages/hatake_yaml/test/conformance_scenario_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:hatake_core/hatake_core.dart';
import 'package:hatake_yaml/hatake_yaml.dart';
import 'package:test/test.dart';

/// Runs the shared scenario fixture (spec/conformance/scenario.json) — the same
/// contract the TypeScript edition (`hatake run`) implements.
///
/// ここで固定するのは「定義を動かした答えが、道具と画面で同じになる」こと。答えの
/// 作り方の順番（normalize → computed → 状態 → 検証）がどちらかでズレると、道具の
/// 答えが嘘になる（AI はその答えを信じて直すので、嘘は一番高くつく）。
void main() {
  final fixture = jsonDecode(
    File('../../../spec/conformance/scenario.json').readAsStringSync(),
  ) as Map<String, dynamic>;

  // strict で読む＝フィクスチャが「本当に書ける定義」であることも縛る。
  final page = parsePageJson(jsonEncode(fixture['page']), strict: true);
  final runner = ScenarioRunner();

  group('conformance: scenario', () {
    for (final raw in fixture['cases'] as List) {
      final map = (raw as Map).cast<String, Object?>();
      final one = ScenarioCase.fromMap(map);
      test(one.name, () {
        final answer = runner.runCase(page, one);
        expect(
          compareAnswer(one.expect, answer),
          isEmpty,
          reason: '答えが表と違います: '
              'errors=${answer.errors.map((e) => '${e.field}=${e.message}')} '
              'computed=${answer.computed} enabled=${answer.enabled} '
              'hidden=${answer.hidden} required=${answer.required}',
        );
      });
    }

    // 期待は「書いた欄だけ」見るので、まるごと同じかもここで見る（表の側が欄を
    // 書き忘れても通ってしまうのを防ぐ）。
    test('答えの中身そのものが表と同じ', () {
      for (final raw in fixture['cases'] as List) {
        final map = (raw as Map).cast<String, Object?>();
        final one = ScenarioCase.fromMap(map);
        final expected = (map['expect']! as Map).cast<String, Object?>();
        final answer = runner.runCase(page, one);

        expect(
          answer.errors.map((e) => '${e.field}=${e.message}').toSet(),
          {
            for (final e in expected['errors']! as List)
              '${(e as Map)['field']}=${e['message']}',
          },
          reason: one.name,
        );
        expect(answer.hidden.toSet(),
            {...(expected['hidden']! as List).map((one) => '$one')},
            reason: one.name);
        expect(answer.required.toSet(),
            {...(expected['required']! as List).map((one) => '$one')},
            reason: one.name);
        expect(answer.enabled, (expected['enabled']! as Map).cast<String, bool>(),
            reason: one.name);
        (expected['computed']! as Map).forEach((key, value) {
          final actual = answer.computed['$key'];
          // JSON から読んだ数と計算した数は int / double が混ざる（1150 と 1150.0）。
          // 画面に出るのは同じ数なので、数として比べる。
          if (value is num && actual is num) {
            expect(actual.toDouble(), value.toDouble(), reason: '$key（${one.name}）');
          } else {
            expect(actual, value, reason: '$key（${one.name}）');
          }
        });
      }
    });
  });
}
